code/week6/main.cpp

```diff
@@ -1,93 +1,104 @@
-
-    // iterate over an image the safe way
-    for(auto it = my_image.begin<cv::Vec3b>(); it != my_image.end<cv::Vec3b>(); it++) {
-        std::cout << "Blue = " << std::to_string((*it)[0]) << std::endl;
-        std::cout << "Green = " << std::to_string((*it)[1]) << std::endl;
-        std::cout << "Red = " << std::to_string((*it)[2]) << std::endl;
-    }
-    std::cout << "\n";
+using namespace std;
 
-    // faster way to iterate over an image
-    int channels = my_image.channels();
-    std::cout << "my_image has " << channels << " channels" << std::endl;
-    int nRows = my_image.rows;
-    std::cout << "my_image has " << nRows << " rows" << std::endl;
-    std::cout << "my_image has " << my_image.cols << " cols" << std::endl;
-    // here we multiply by the number of channels since we will be dereferencing it with uchar
-    // and each channel is a uchar
-    int nCols = my_image.cols * channels;
+void generateComputerMove(const Board &board, unsigned int &r_out, unsigned int &c_out) {
+  static random_device rand_dev;
+  static default_random_engine engine(rand_dev());
+  static uniform_int_distribution<unsigned int> uniform_dist(0,2);
+  do {
+    r_out = uniform_dist(engine);
+    c_out = uniform_dist(engine);
+  } while(board.getMarker(r_out, c_out) != Marker::Empty);
+}
 
-    // if the image is stored continuously then we have only a single row of
-    // length nRows * nCols
-    if(my_image.isContinuous()) {
-        nCols *= nRows;
-        nRows = 1;
+void getMoveFromPlayer(unsigned int &r_out, unsigned int &c_out) {
+  static regex coordinate_patter("[0-3],[0-3]");
+  do {
+    cout << "Your move (r,c): ";
+    string input;
+    cin >> input;
+    if(regex_match(input,coordinate_patter)) {
+      auto comma_loc = input.find(',');
+      r_out = static_cast<unsigned int>(stoi(input.substr(0,comma_loc)));
+      c_out = static_cast<unsigned int>(stoi(input.substr(comma_loc+1)));
+      return;
+    } else {
+      cout << "Sorry, I couldn't understand your coordinates!" << endl;
     }
+  } while(true);
+}
 
-    for(int i = 0; i < nRows; i++) {
-        // get pointer to the start of the current row
-        uchar* current = my_image.ptr<uchar>(i);
-        for(int j = 0; j < nCols; j++) {
-            // prints out the value at the nCols in nRows
-            std::cout << std::to_string(current[j]);
-            // formats the printing of the statement
-            if((j + 1) % 3 != 0) {
-                std::cout << ",";
-            } else {
-                std::cout << std::endl;
-            }
-        }
-    }
+void printBoard(const Board &board) {
+  auto charForMarker = [](const Marker &m) {
+      switch(m) {
+        case Marker::Empty:
+          return ' ';
+        case Marker::X:
+          return 'X';
+        case Marker::O:
+          return 'O';
+      }
+  };
+  std::string layout = " | | \n-----\n | | \n-----\n | | ";
+  layout[0] = charForMarker(board.getMarker(0,0));
```

Comment:
Hard coding these positions smells bad, why not use printf instead (so this can be readable)?

IMO, this should probably be properly printed out in a loop, this approach is really hacky.

Reply:
That's fair. I was being lazy